### PR TITLE
Ability to write mermaid diagrams via code blocks (like gitHub)

### PR DESCRIPTION
**Pull request by nkuehn**
implements #1251 

 - [X] the styling and theming is already pretty acceptable to start with. Design should look over the preview site and recommend incremental changes but I would recommend against starting a design process from scratch because it's very hard to describe what can be customized in mermaid and what not. 
 - [X] it does generally work, the visual integration test is sufficient, the author experience is also OK (syntax help and highlighting in VSCode, live update when editing, experience identical to GitHub)
 - [X]  the way it renders is as proper as it gets with mermaid.  It's not starting the DOM crawl like the normal mermaid behavior but renders through the mermaid API when the props change. Can't get around `__dangerouslySetInnerHTML` by design
 - [X] documentation
 - [x] get feedback from design on the style and color choices. 

This voluntarily is a client side rendering architecture because that is what mermaid is built, optimized, documented and tested against.  Server side prerendering is technically possible and some plugins do it but it's slow and it's not possible to use links or other interactivity in the diagrams.  The latter is a blocker to be so I suggest to stick with client side. 

I did a check with the bundle size analyzer.  The mermaid package is pretty large (260kb gzipped) but it's loaded lazy client side  (React.Lazy & suspense) so it's a bundle on its own that is not in the critical path to site interactivity.

### CODE REVIEW

**Comment by nkuehn** [comment]
@Shecki  could you be so nice and take a look at the color choices for the diagrams the docs kit can generate now:

 * https://docskit-pr-1253.commercetools.vercel.app/docs-smoke-test/components/mermaid-diagram  
 * https://docskit-pr-1253.commercetools.vercel.app/documentation/writing/images#mermaid-diagrams

We picked from the available color tokens whereever possible.  Just the yellow info boxes had no well matching predefined color, they are just browser standard "yellow" and "lightyellow" now.  Info notes are conventionally yellow in diagrams, but if you propose another color we can use that, too. 

To me it's good enough.  The colors can be changed centrally later for the whole website (the diagrams are generated), so there's no pressure to have it 100% perfect immediately.

File: packages/gatsby-theme-docs/package.json [hunk 1]
```diff
@@ -76,7 +77,13 @@
...
     "rehype-slug": "4.0.1",
     "remark-emoji": "2.2.0",
-    "slugify": "1.6.5"
+    "remark-frontmatter": "3.0.0",
+    "remark-parse": "9.0.0",
+    "remark-rehype": "8.1.0",
+    "slugify": "1.6.5",
+    "unified": "9.2.2",
+    "unist-util-filter": "2.0.3",
+    "unist-util-visit": "2.0.3"
```
Comment (nkuehn) [inline packages/gatsby-theme-docs/package.json:86]: TBD understand why these are all added, only unist-util-visit is used in the new code.  I think it's a fix for warnings that should have been done anyways but maybe this is rather an artifact or bug.

File: packages/gatsby-theme-docs/src/plugins/remark-mdx-mermaid.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+const visit = require('unist-util-visit');
+
+/**
+ * Given the MDXAST ast, look for all fenced codeblocks that have a language of
```
Comment (nkuehn) [inline packages/gatsby-theme-docs/src/plugins/remark-mdx-mermaid.js:4]: This design choice is here for two reasons:
 - people will get used to being able to use fenced code blocks for mermaid diagrams in GitHub
 - you get syntax highlighting in the editor if it's a fenced code block but not if the syntax would be the body content of an <Mermaid> tag in MDX

File: packages/gatsby-theme-docs/src/plugins/remark-mdx-mermaid.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * MDXAST reference: https://github.com/mdx-js/specification#jsx-1
+ *
+ * @param {object}  ast
+ * @return {function}
+ */
+module.exports = () => (ast) => {
+  visit(ast, { type: 'code', lang: 'mermaid' }, (node) => {
+    return Object.assign(node, {
+      type: 'jsx',
+      value: '<Mermaid graph={`' + node.value + '`} />',
```
Comment (nkuehn) [inline packages/gatsby-theme-docs/src/plugins/remark-mdx-mermaid.js:15]: Passing the graph as children body content instead of as a prop caused various encoding and robustness issues. so it's passed as a prop now.

File: packages/ui-kit/src/components/mermaid-client-side.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  display: flex;
+  justify-content: center;
+  line-height: normal;
+  a span.nodeLabel {
+    color: ${colors.light.link} !important;
+    text-decoration: underline;
+  }
+`;
+
+const idForGraph = (graph: string) => `mermaid-${murmurhash.v3(graph)}`;
```
Comment (nkuehn) [inline packages/ui-kit/src/components/mermaid-client-side.tsx:111]: murmurhash is a compact hash implementation that is optimized for being collision free.  The package is smaller than the wrappers around nodejs crypto APIs that could have provided md5 or sha and it's the right hash for the use case (mermaid needs a unique ID for every graph to not get confused)

File: packages/ui-kit/src/components/mermaid.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import LoadingSpinner from '@commercetools-uikit/loading-spinner';
+
+// This is a client-side only implementation of mermaid, which is its primary development target.
+// https://www.gatsbyjs.com/docs/using-client-side-only-packages/
+// Server side rendering mermaid diagrams is technically possible
+// and some remark and gatsby plugins do it. But all are badly maintained
+// and the server side rendering use case is not covered in the mermaid documentation,
+// which makes it a hard to maintain architecture.
+// server side rendering also has the disadvantage that clickable elements cannot
+// work because this requires attaching event handlers.
```
Comment (nkuehn) [inline packages/ui-kit/src/components/mermaid.tsx:11]: We're not using this but its an interesting outlook to interactivity with the graphs. Not used for now.

File: packages/ui-kit/src/components/mermaid.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+const MermaidLazy = React.lazy(() => import('./mermaid-client-side'));
```
Comment (nkuehn) [inline packages/ui-kit/src/components/mermaid.tsx:13]: There are other choices for client-side only components in gatsbyJS, too.  React Lazy+Suspense is chosen because it does not incur extra libraries overhead, works and also provides lazy loaded bundle splitting for the big mermaidJS library.

File: recommended.code-workspace [hunk 7]
```diff
@@ -63,7 +63,9 @@
       "redhat.vscode-yaml",
       "silvenon.mdx",
       "jounqin.vscode-mdx",
-      "errata-ai.vale-server"
+      "errata-ai.vale-server",
+      "bierner.markdown-mermaid",
+      "bpruitt-goddard.mermaid-markdown-syntax-highlighting"
```
Comment (nkuehn) [inline recommended.code-workspace:68]: When applying this to the content repo, please also add these to the recommended plugins in the content repo.

File: recommended.code-workspace [hunk 8]
```diff
@@ -76,7 +78,7 @@
           "<node_internals>/**"
         ],
         "program": "${workspaceRoot}/node_modules/.bin/gatsby",
-        "cwd": "${workspaceFolder}/websites/docs-smoke-test",
+        "cwd": "${workspaceRoot}/websites/docs-smoke-test",
```
Comment (nkuehn) [inline recommended.code-workspace:81]: unrelated fix that is not worth an own PR.

File: websites/docs-smoke-test/src/content/components/mermaid-diagram.mdx [hunk 9]
```diff
@@ -0,0 +1,85 @@
+---
+title: Mermaid Diagrams
+---
+
+It's possible to generate Mermaid Diagrams by either placing the Mermaid code into `<Mermaid>...</Mermaid>` components or by placing a code block with language `mermaid`.
```
Comment (nkuehn) [inline websites/docs-smoke-test/src/content/components/mermaid-diagram.mdx:5]: TODO this is not correct any more.  Remove the sentence
```suggestion
```

File: websites/documentation/src/content/index.mdx [hunk 10]
```diff
@@ -1,5 +1,43 @@
...
+- Custom site-specific components can be centrally injected into MDX (for example interactive features)
+- [Dynamic API documentation](/api-docs/raml) from RAML API specifications (using rmf-codegen for parsing and evaluation)
+  - "hybrid" API documentation mode: Full control on what page which endpoint and type is documented in which order, but the actual endpoint and type specs are fully generated. Free form MDX content can be interleaved with API documentation.
+  - Resources and Types are autolinked no matter where they are placed in the site. No manual maintenance of API documentation links when refactoring the site structure.
+  - scales into very large APIs.
+
+The docs kit follows the convention over configuration principle, especially concerning file naming, file locations and navigation structure. This makes it easy for an author to contribute to any existing documentation website. The kit is [configurable](/configuration/setup) nevertheless.
+
+Many components are provided in the standalone `@commercetools-docs/ui-kit` NPM package and can be used in other website that want to take over the basic design language of the documentation.
```
Comment (nkuehn) [inline websites/documentation/src/content/index.mdx:35]: Unrelated to mermaid.  I took the chance to give our poor homepage a little feature marketing.

File: websites/documentation/src/content/writing/images.mdx [hunk 11]
```diff
@@ -1,17 +1,67 @@
 ---
-title: Images
+title: Images and Diagrams
 wideLayout: true
 ---
 
-You can find all images in `/src/images`, and can use standard Markdown syntax to insert images. It is good to have a meaningful image names and alt text when using images. A meaningful alt text is useful for visually impaired users using softwares to read text on a page. Also, a meaningful image name with keywords helps with SEO.
+# Images
```
Comment (nkuehn) [inline websites/documentation/src/content/writing/images.mdx:6]: I chose to document the diagram feature on the same page like images and wrote a more proper documentation for images on that occasion.  It was pretty poor.

File: websites/documentation/src/content/writing/images.mdx [hunk 12]
```diff
@@ -1,17 +1,67 @@
...
+![a large dog](/images/dog.jpg 'This dog serves as an example image and this is the title text')
+
+</SideBySide>
+
+Unless not otherwise possible, avoid:
+
+- images on external servers (not optimized, can disappear)
+- images of small sizes or large heights
+
+# Mermaid Diagrams
```
Comment (nkuehn) [inline websites/documentation/src/content/writing/images.mdx:39]: The actual docs.  Not much to explain, we should not even start explaining the syntax here but only refer to the mermaid docs.

File: websites/documentation/src/content/writing/images.mdx [hunk 13]
```diff
@@ -1,17 +1,67 @@
...
+- You can add a caption to images by following the Markdown syntax for titles.
+- Images with a transparent background (PNG) are rendered on white background. It's OK to leave transparent background
+- SVG images are not optimized or scaled, but would still be rendered. They need to be placed into `/src/files/` for unprocessed delivery.
+
+<SideBySide>
+
+```markdown
+Full example with alt text title:
+
+![a large dog](/images/dog.jpg 'This dog serves as an example image and this is the title text')
```
Comment (nkuehn) [inline websites/documentation/src/content/writing/images.mdx:23]: The dog image is under a creative commons license.

**Review by nkuehn — COMMENTED** [review]
explanatory comments

File: packages/gatsby-theme-docs/package.json (hunk 1 above)
Comment (nkuehn) [reply packages/gatsby-theme-docs/package.json:86]: fixed, the dependencies were forgotten to be removed.

**Review by timonrey — APPROVED** [review]
Implementation looks good. Thanks for spending some time on the documentation! 🙌🏼 💯 
We still need an approval for the design 💅🏼

**Comment by Shecki** [comment]
@timonrey @nkuehn 
colors are fine for me
